### PR TITLE
Refactor code for testing; add some unit and integration tests.

### PR DESCRIPTION
**Pull request by brikis98**
This PR does the following:
1. Refactor the code from two huge functions, one in `create_command` and one in `delete_command` into smaller functions that are more readable and more testable.
2. Add a basic end-to-end integration test that creates an EC2 instance, uses `create_command` to take a snapshot of it, and then uses `delete_command` to delete that snapshot. 
3. Add basic unit tests for the date parsing. However, this should be refactored to use `time.ParseDuration` in the future.
4. Adds a `circle.yml` file to run these tests in CI. @josh-padnick, could you enable a Circle CI build for this repo and add AWS credentials as environment variables? I don’t have access to it, so I can’t do it myself. 
5. Control aws region through a command line param rather than environment variables. This makes testing easier. It also allows you to run the `--help` command without getting an error about a missing environment variable.

There are a lot more unit and integration tests that should be added, especially for the deletion pieces, but I’m trying to strike a balance between making this code pristine via testing vs using this code in the real world (with clients). Hopefully, this lays down a decent base for tests, and we can add on more on top of it.

### CODE REVIEW

File: circle.yml [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+    # Get our repo into the CircleCI GOPATH.
+    - |
+        mkdir -p ${REPO}
+        rmdir ${REPO}
+        ln -s "/home/ubuntu/${CIRCLE_PROJECT_REPONAME}" "${REPO}"
+
+test:
+  override:
+    - |
+        ${REPO}/_ci/run-tests.sh
```
Comment (josh-padnick) [inline circle.yml:18]: Will this download the needed go dependencies? I believe aws sdk for golang is the only one needed in this case.

File: create_command.go [hunk 2]
```diff
@@ -56,19 +61,32 @@ func (c *CreateCommand) Run(args []string) int {
...
 	// Check for required command-line args
+	if c.AwsRegion == "" {
+		return snapshotId, errors.New("ERROR: The argument '--region' is required.")
+	}
+
 	if c.InstanceId == "" {
-		c.Ui.Error("ERROR: The argument '--instance' is required.")
-		return 1
+		return snapshotId, errors.New("ERROR: The argument '--instance' is required.")
 	}
```
Comment (josh-padnick) [inline create_command.go:82]: Good call. I wrote this as a golang newb and didn't know better at the time. :-)

File: delete_command.go [hunk 3]
```diff
@@ -59,174 +66,243 @@ func (c *DeleteCommand) Run(args []string) int {
...
-	} else if err != nil {
-		panic(err)
-	}
-	if len(resp.Images) == 0 {
-		c.Ui.Error("No AMIs were found for EC2 instance \"" + c.InstanceId + "\"")
-		return 0
+	images, err := findImages(c.InstanceId, svc)
+	if err != nil {
+		return err
 	}
```
Comment (josh-padnick) [inline delete_command.go:93]: This refactor is much cleaner and more readable.

File: integration_create_delete_test.go [hunk 4]
```diff
@@ -0,0 +1,206 @@
...
+	if err != nil {
+		t.Fatal(err)
+	}
+}
+
+func waitForInstanceToStart(instance *ec2.Instance, svc *ec2.EC2, logger *log.Logger, t *testing.T) {
+	logger.Printf("Waiting for instance %s to start...", *instance.InstanceId)
+	if err := svc.WaitUntilInstanceRunning(&ec2.DescribeInstancesInput{InstanceIds: []*string{instance.InstanceId}}); err != nil {
+		t.Fatal(err)
+	}
```
Comment (josh-padnick) [inline integration_create_delete_test.go:108]: Nice find in the AWS SDK on `WaitUntilInstanceRunning`!

File: unit_parse_hours_test.go [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+
+func TestParseOlderThanToHoursTenDays(t *testing.T) {
+	t.Parallel()
+	testParseOlderThan("10d", 240, t)
+}
+
+func TestParseOlderThanToHoursNineHundredNinetyNineDays(t *testing.T) {
+	t.Parallel()
+	testParseOlderThan("999d", 23976, t)
+}
```
Comment (josh-padnick) [inline unit_parse_hours_test.go:90]: Nice style of tests for the date parser.

**Comment by josh-padnick** [comment]
I have one comment on the `circle.yml` but otherwise these changes are a big improvement. This was my first golang program and a side project so original code quality was mediocre; this makes it much better!

I've added the requested creds to CircleCI and will grant you write access to this repo as well so that you can merge the PR yourself once you validate your `circle.yml`. Thanks for this!

File: circle.yml (hunk 1 above)
Comment (brikis98) [inline circle.yml:18]: I don't want to switch over to glide for one dependency, so I tried to just run `go test -i`. https://github.com/brikis98/ec2-snapper/commit/51c8f761db78037038d590f0d5a976b85ad6754c

File: integration_create_delete_test.go (hunk 4 above)
Comment (brikis98) [inline integration_create_delete_test.go:108]: Heh, the irony is I actually wrote my own version that kept checking the instance status in a loop, with sleep in between... And then while writing some code later on, I stumbled across that function in IntelliJ's auto complete :)

**Comment by brikis98** [comment]
Ah, so this is interesting. When you run this build in Circle, it says:

```
Suppressing export of environment variables AWS_ACCESS_KEY_ID, AWS_SECRET_ACCESS_KEY fork PR builds
```

This makes sense, as you don't want to expose secrets to random folks who submit a forked PR. However, it also means the integration tests will fail on forks. I'm not sure there is any great solution for now, so I'm going to merge this in and see if the tests pass.

File: circle.yml (hunk 1 above)
Comment (josh-padnick) [inline circle.yml:18]: Agreed we don't need glide for a single dependency, but good to see you added `go test -i`.

**Comment by josh-padnick** [comment]
> However, it also means the integration tests will fail on forks.

I was thinking about this while adding creds. Could someone exploit this to launch arbitrary resources in our AWS account with this? Then I realized Circle will only run for branches in the official repo, not forks, so we get to be gatekeepers of any such changes.

I didn't realize Circle will still run on forked branches but specifically suppress AWS env vars, which it must specially recognize as secret. Thoughtful on their part.

**Comment by brikis98** [comment]
> I didn't realize Circle will still run on forked branches but specifically suppress AWS env vars, which it must specially recognize as secret.

They suppress all env vars, actually. Which is very smart given how easily someone could steal them with an arbitrary PR :)

**Pull request merged**